File: python/generate4BillionIPs.py
# Python script version of 4BillionIPs written by @mukeenanyafiq

import time
IP_COUNT = 2**32  # 4 billion IPs
BATCH_SIZE = 1000000  # Write 1 million IPs at a time
LOG_INTERVAL = IP_COUNT // 100  # Log progress every 1% completion
filename = "../4BillionIPs.txt"

def int_to_ip(int_ip):
    return f"{(int_ip >> 24) & 0xFF}.{(int_ip >> 16) & 0xFF}.{(int_ip >> 8) & 0xFF}.{int_ip & 0xFF}"

start = time.time()  # Record execution time

with open(filename, 'w') as file:
    buffer = []

    for i in range(IP_COUNT):
        buffer.append(int_to_ip(i))

        # If buffer size reaches BATCH_SIZE, write to file
        if len(buffer) == BATCH_SIZE:
            file.write("\n".join(buffer) + "\n")
            buffer = []  # Clear the buffer

        # Log progress at intervals
        if i > 0 and i % LOG_INTERVAL == 0: print(f"[Progress] Generated {i:,}/{IP_COUNT:,} IPs ({(i / IP_COUNT) * 100:.2f}%)")

    # Write any remaining IPs in the buffer to the file
    if buffer: file.write("\n".join(buffer) + "\n")

print(f"IP generation complete and saved to file. ({time.time() - start:.2f}s)")

File: generate4BillionIPs.js
const fs = require('fs');
const IP_COUNT = Math.pow(2, 32) + 1;  // 4 billion IPs
const BATCH_SIZE = 1000000;            // Write 1 million IPs at a time
const LOG_INTERVAL = Math.floor(IP_COUNT / 100); // Log progress every 1% completion

function intToIP(int) {
    return [
        (int >>> 24) & 0xFF,
        (int >>> 16) & 0xFF,
        (int >>> 8) & 0xFF,
        int & 0xFF
    ].join('.');
}

let buffer = [];
for (let i = 0; i < IP_COUNT; i++) {
    buffer.push(intToIP(i));

    // If buffer size reaches BATCH_SIZE, write to file
    if (buffer.length == BATCH_SIZE) {
        fs.appendFileSync("./4BillionIPs.txt", buffer.join("\n") + "\n");
        buffer = [];  // Clear the buffer
    }

    // Log progress at intervals
    if (i > 0 && i % LOG_INTERVAL == 0) console.log(`[Progress] Generated ${i} / ${IP_COUNT} IPs (${(i / IP_COUNT) * 100}%)`);
}

// Write any remaining IPs in the buffer to the file
if (buffer.length > 0) fs.appendFileSync("./4BillionIPs.txt", buffer.join("\n") + "\n");
console.log("IP generation complete and saved to file.");